File: native/android/voice/sakha/SakhaPauseParser.kt
/**
 * Sakha Pause Parser
 *
 * Streaming parser for the Sakha persona's text protocol. Sakha emits prose
 * with three pause markers — <pause:short>, <pause:medium>, <pause:long> —
 * and an inline FILTER_FAIL sentinel when the Wisdom Core cannot retrieve a
 * verse honestly tied to what the user said.
 *
 * Design constraints:
 * - Must accept partial chunks. The SSE stream may split a marker across
 *   frames (e.g. "<pau" + "se:short>"). The parser buffers an incomplete
 *   tail and re-tries on the next chunk.
 * - Must emit speakable segments greedily so the TTS player can start
 *   speaking as soon as the first natural break (sentence end, comma, or
 *   pause marker) arrives. This is the latency win over "wait for full
 *   response then synthesize."
 * - Must distinguish Sanskrit segments (Devanagari script) from prose so
 *   the player can route them to a slower, reverent voice.
 *
 * Output:
 * - [PauseEvent.Speak]    — text ready to be sent to TTS
 * - [PauseEvent.Pause]    — honour a quiet beat
 * - [PauseEvent.Filter]   — FILTER_FAIL detected; abort speakable output
 *
 * The parser is deliberately stateful and single-threaded. Wrap calls in a
 * mutex if you share an instance across threads.
 */

package com.mindvibe.kiaan.voice.sakha

sealed class PauseEvent {
    data class Speak(val text: String, val isSanskrit: Boolean) : PauseEvent()
    data class Pause(val durationMs: Long) : PauseEvent()
    object Filter : PauseEvent()
}

class SakhaPauseParser {

    companion object {
        private const val PAUSE_SHORT_MS = 300L
        private const val PAUSE_MEDIUM_MS = 600L
        private const val PAUSE_LONG_MS = 1200L

        // Conservative Devanagari Unicode range. Covers Sanskrit + Hindi
        // verses; we treat any segment that contains Devanagari as Sanskrit
        // for TTS routing purposes (Sarvam handles Hindi via the same voice
        // family but with prose prosody, which is fine).
        private val DEVANAGARI_REGEX = Regex("""[ऀ-ॿ]""")

        // The longest token a partial chunk could *not* yet form. We use
        // this to decide when to flush vs keep buffering.
        private const val LONGEST_PARTIAL_MARKER = "<pause:medium>".length

        // Natural sentence-end characters across English + Indian scripts.
        // We add Devanagari danda (।) and double-danda (॥) for Sanskrit
        // verse halves.
        private const val SENTENCE_ENDS = ".!?।॥"

        // FILTER_FAIL sentinel — must match the persona spec exactly.
        private const val FILTER_FAIL_TOKEN = "FILTER_FAIL: no_retrieval"
    }

    private val buffer = StringBuilder()
    private var filterFailEmitted = false

    /**
     * Feed a streaming text chunk. Returns the events that can be safely
     * emitted right now; any incomplete tail stays buffered.
     */
    fun feed(chunk: String): List<PauseEvent> {
        if (filterFailEmitted) return emptyList()
        if (chunk.isEmpty()) return emptyList()
        buffer.append(chunk)
        return drain(force = false)
    }

    /**
     * Signal end-of-stream. Flushes any buffered prose as a final Speak event
     * (without waiting for a sentence boundary).
     */
    fun finish(): List<PauseEvent> {
        if (filterFailEmitted) return emptyList()
        return drain(force = true)
    }

    /** Reset for the next turn. */
    fun reset() {
        buffer.setLength(0)
        filterFailEmitted = false
    }

    // ------------------------------------------------------------------------
    // Internals
    // ------------------------------------------------------------------------

    private fun drain(force: Boolean): List<PauseEvent> {
        val events = mutableListOf<PauseEvent>()

        // Detect FILTER_FAIL anywhere in the buffer — this is a hard stop.
        val ffIdx = buffer.indexOf(FILTER_FAIL_TOKEN)
        if (ffIdx >= 0) {
            // Drop anything before and after; the persona spec says the
            // server falls back to a template tier on this signal, so we
            // do NOT speak whatever leaked before the sentinel.
            buffer.setLength(0)
            filterFailEmitted = true
            events.add(PauseEvent.Filter)
            return events
        }

        while (true) {
            // Look for the next pause marker.
            val markerStart = buffer.indexOf('<')

            if (markerStart < 0) {
                // No marker in sight. Flush a speakable slice if we can.
                val flushed = flushSpeakable(force) ?: break
                events.add(flushed)
                continue
            }

            // There's a '<' somewhere. Try to parse a marker starting there.
            val parsed = parsePauseMarker(markerStart)
            if (parsed == null) {
                // No valid pause marker at this position. Two cases:
                //
                // 1) The `<` could be the start of an *incomplete* marker
                //    that the next chunk will complete (e.g. "<pau"). We
                //    detect this by: not forcing AND no `>` yet AND the
                //    tail is short enough to plausibly become a marker.
                //    In that case, flush prose before the `<` and keep the
                //    tail buffered for the next feed.
                //
                // 2) Otherwise the `<` is part of a different (unknown)
                //    tag, e.g. "<not-a-pause>". The model is only supposed
                //    to emit pause markers, so any other tag is dropped
                //    cleanly: we keep the prose around it and skip past
                //    the `>`. We do NOT speak "<", "not-a-pause", or ">"
                //    as literal text — that would mangle TTS prosody.
                val closeIdx = buffer.indexOf('>', markerStart)
                if (!force && closeIdx < 0 && (buffer.length - markerStart) <= LONGEST_PARTIAL_MARKER) {
                    if (markerStart > 0) {
                        val pre = buffer.substring(0, markerStart)
                        buffer.delete(0, markerStart)
                        emitSpeakable(events, pre)
                    }
                    break
                }

                // Drop the unknown tag inline; preserve surrounding prose.
                if (closeIdx < 0) {
                    // Forced flush with no closer ever arrived — treat the
                    // `<` and trailing junk as literal text rather than
                    // dropping it silently.
                    val pre = buffer.toString()
                    buffer.setLength(0)
                    emitSpeakable(events, pre)
                    break
                }
                // Splice prose around the unknown tag together.
                val before = buffer.substring(0, markerStart)
                val after = buffer.substring(closeIdx + 1)
                buffer.setLength(0)
                buffer.append(before)
                buffer.append(after)
                continue
            }

            val (durationMs, markerEnd) = parsed
            // Flush prose before the marker first…
            if (markerStart > 0) {
                val pre = buffer.substring(0, markerStart)
                buffer.delete(0, markerStart)
                emitSpeakable(events, pre)
                // After the delete, the marker now sits at index 0.
                val newEnd = markerEnd - markerStart
                buffer.delete(0, newEnd)
            } else {
                buffer.delete(0, markerEnd)
            }
            events.add(PauseEvent.Pause(durationMs))
        }

        return events
    }

    /**
     * Try to parse a pause marker starting at [start]. Returns
     * (durationMs, indexAfterMarker) on success, null if the marker is
     * incomplete or malformed.
     */
    private fun parsePauseMarker(start: Int): Pair<Long, Int>? {
        val close = buffer.indexOf('>', start)
        if (close < 0) return null
        val token = buffer.substring(start, close + 1)
        val durationMs = when (token.lowercase()) {
            "<pause:short>" -> PAUSE_SHORT_MS
            "<pause:medium>" -> PAUSE_MEDIUM_MS
            "<pause:long>" -> PAUSE_LONG_MS
            else -> return null  // Some other tag — treat as literal in caller.
        }
        return durationMs to (close + 1)
    }

    /**
     * Try to flush a speakable slice from the head of the buffer.
     *
     * Strategy:
     *  - If [force]: flush whatever is buffered as one final segment.
     *  - Otherwise: only flush up to the last sentence-ending punctuation,
     *    because Sarvam pacing is dramatically better with whole sentences
     *    than with mid-sentence fragments.
     */
    private fun flushSpeakable(force: Boolean): PauseEvent.Speak? {
        if (buffer.isEmpty()) return null

        val sliceEnd: Int = if (force) {
            buffer.length
        } else {
            findLastSentenceBoundary() ?: return null
        }

        val raw = buffer.substring(0, sliceEnd)
        buffer.delete(0, sliceEnd)
        val cleaned = raw.trim()
        if (cleaned.isEmpty()) return null
        return PauseEvent.Speak(cleaned, containsDevanagari(cleaned))
    }

    private fun findLastSentenceBoundary(): Int? {
        var lastBoundary = -1
        for (i in buffer.indices) {
            if (SENTENCE_ENDS.indexOf(buffer[i]) >= 0) {
                lastBoundary = i
            }
        }
        return if (lastBoundary >= 0) lastBoundary + 1 else null
    }

    private fun emitSpeakable(events: MutableList<PauseEvent>, raw: String) {
        val cleaned = raw.trim()
        if (cleaned.isEmpty()) return
        events.add(PauseEvent.Speak(cleaned, containsDevanagari(cleaned)))
    }

    private fun containsDevanagari(text: String): Boolean {
        return DEVANAGARI_REGEX.containsMatchIn(text)
    }
}

File: kiaanverse-mobile/apps/mobile/native/ios/Sources/Sakha/SakhaPauseParser.swift
/**
 * Sakha Pause Parser (iOS port)
 *
 * Direct port of apps/mobile/native/android/.../SakhaPauseParser.kt. Pure
 * single-threaded streaming parser. The output PauseEvent is a struct
 * defined in SakhaTypes.swift so SakhaVoiceManager + SakhaTtsPlayer share
 * a single type — same as Android.
 *
 * Behaviour:
 * - Accepts partial chunks: an SSE frame like "<pau" + "se:short>" buffers
 *   the incomplete tail and flushes once the next frame completes the
 *   marker.
 * - Greedy speakable flush at sentence ends so TTS can start as early as
 *   the first natural break.
 * - Detects Devanagari (the Unicode block U+0900–U+097F) so the player
 *   routes Sanskrit segments to the reverent voice (sanskritVoiceId).
 * - FILTER_FAIL sentinel is hard-stop — drops everything before/after.
 */

import Foundation

public final class SakhaPauseParser {

    private static let pauseShortMs: Int64 = 300
    private static let pauseMediumMs: Int64 = 600
    private static let pauseLongMs: Int64 = 1200

    /// Conservative Devanagari range — Sanskrit + Hindi.
    private static let devanagariRegex = try! NSRegularExpression(pattern: "[\\u0900-\\u097F]")

    /// Longest token a partial chunk could not yet form. Keep buffer < this
    /// to decide between "wait for next frame" and "treat as unknown tag".
    private static let longestPartialMarker: Int = "<pause:medium>".count

    /// Sentence-end characters across English + Indian scripts.
    /// Includes the Devanagari danda (U+0964) and double-danda (U+0965).
    private static let sentenceEnds: Set<Character> = [".", "!", "?", "।", "॥"]

    /// FILTER_FAIL sentinel — must match the persona spec exactly.
    private static let filterFailToken = "FILTER_FAIL: no_retrieval"

    private var buffer = ""
    private var filterFailEmitted = false

    public init() {}

    /// Feed a streaming text chunk. Returns the events that can be safely
    /// emitted right now; any incomplete tail stays buffered.
    public func feed(_ chunk: String) -> [PauseEvent] {
        if filterFailEmitted { return [] }
        if chunk.isEmpty { return [] }
        buffer.append(chunk)
        return drain(force: false)
    }

    /// Signal end-of-stream. Flushes any buffered prose as a final Speak.
    public func finish() -> [PauseEvent] {
        if filterFailEmitted { return [] }
        return drain(force: true)
    }

    public func reset() {
        buffer = ""
        filterFailEmitted = false
    }

    // MARK: - Internals

    private func drain(force: Bool) -> [PauseEvent] {
        var events: [PauseEvent] = []

        // Hard stop on FILTER_FAIL.
        if let range = buffer.range(of: Self.filterFailToken) {
            _ = range
            buffer = ""
            filterFailEmitted = true
            events.append(.filter)
            return events
        }

        while true {
            guard let markerStart = buffer.firstIndex(of: "<") else {
                if let flushed = flushSpeakable(force: force) {
                    events.append(flushed)
                    continue
                }
                break
            }

            let parsed = parsePauseMarker(at: markerStart)
            if parsed == nil {
                // Could be incomplete: wait for next chunk.
                let closeIdx = buffer.range(of: ">", range: markerStart..<buffer.endIndex)?.lowerBound
                let tailLen = buffer.distance(from: markerStart, to: buffer.endIndex)
                if !force && closeIdx == nil && tailLen <= Self.longestPartialMarker {
                    if markerStart > buffer.startIndex {
                        let pre = String(buffer[buffer.startIndex..<markerStart])
                        buffer.removeSubrange(buffer.startIndex..<markerStart)
                        emitSpeakable(into: &events, raw: pre)
                    }
                    break
                }
                // Drop unknown tag inline; preserve surrounding prose.
                guard let closeIdx2 = closeIdx else {
                    // Forced flush, no closer ever — treat tail as literal.
                    let pre = buffer
                    buffer = ""
                    emitSpeakable(into: &events, raw: pre)
                    break
                }
                let before = String(buffer[buffer.startIndex..<markerStart])
                let after = String(buffer[buffer.index(after: closeIdx2)..<buffer.endIndex])
                buffer = before + after
                continue
            }

            let (durationMs, markerEnd) = parsed!
            // Flush prose before the marker.
            if markerStart > buffer.startIndex {
                let pre = String(buffer[buffer.startIndex..<markerStart])
                buffer.removeSubrange(buffer.startIndex..<markerEnd)
                emitSpeakable(into: &events, raw: pre)
            } else {
                buffer.removeSubrange(buffer.startIndex..<markerEnd)
            }
            events.append(.pause(durationMs: durationMs))
        }

        return events
    }

    /// Try to parse a pause marker starting at `start`. Returns (durationMs,
    /// indexAfterMarker) on success, nil if incomplete or malformed.
    private func parsePauseMarker(at start: String.Index) -> (Int64, String.Index)? {
        guard let close = buffer.range(of: ">", range: start..<buffer.endIndex)?.lowerBound else {
            return nil
        }
        let token = String(buffer[start...close]).lowercased()
        let durationMs: Int64
        switch token {
        case "<pause:short>": durationMs = Self.pauseShortMs
        case "<pause:medium>": durationMs = Self.pauseMediumMs
        case "<pause:long>": durationMs = Self.pauseLongMs
        default: return nil
        }
        return (durationMs, buffer.index(after: close))
    }

    private func flushSpeakable(force: Boolean) -> PauseEvent? {
        if buffer.isEmpty { return nil }

        let sliceEnd: String.Index
        if force {
            sliceEnd = buffer.endIndex
        } else if let boundary = lastSentenceBoundary() {
            sliceEnd = boundary
        } else {
            return nil
        }

        let raw = String(buffer[buffer.startIndex..<sliceEnd])
        buffer.removeSubrange(buffer.startIndex..<sliceEnd)
        let cleaned = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if cleaned.isEmpty { return nil }
        return .speak(text: cleaned, isSanskrit: containsDevanagari(cleaned))
    }

    private func lastSentenceBoundary() -> String.Index? {
        var lastBoundary: String.Index? = nil
        var i = buffer.startIndex
        while i < buffer.endIndex {
            if Self.sentenceEnds.contains(buffer[i]) {
                lastBoundary = buffer.index(after: i)
            }
            i = buffer.index(after: i)
        }
        return lastBoundary
    }

    private func emitSpeakable(into events: inout [PauseEvent], raw: String) {
        let cleaned = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if cleaned.isEmpty { return }
        events.append(.speak(text: cleaned, isSanskrit: containsDevanagari(cleaned)))
    }

    private func containsDevanagari(_ text: String) -> Bool {
        let range = NSRange(text.startIndex..., in: text)
        return Self.devanagariRegex.firstMatch(in: text, range: range) != nil
    }
}

/// Swift type alias so `Boolean` reads naturally to Kotlin readers.
private typealias Boolean = Bool
